File: benchmark/c/opl/opl_loadbank_op2/opl_loadbank_op2.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define OPL_EMU_REGISTERS_OPERATORS ( OPL_EMU_REGISTERS_CHANNELS * 2 )
#define OPL_EMU_REGISTERS_WAVEFORMS 8
#define OPL_EMU_REGISTERS_CHANNELS 18
#define OPL_EMU_REGISTERS_REGISTERS 0x200
#define OPL_EMU_REGISTERS_WAVEFORM_LENGTH 0x400

enum opl_emu_envelope_state
{
	OPL_EMU_EG_ATTACK = 1,
	OPL_EMU_EG_DECAY = 2,
	OPL_EMU_EG_SUSTAIN = 3,
	OPL_EMU_EG_RELEASE = 4,
	OPL_EMU_EG_STATES = 6
};
enum op2_flags_t {
  OP2_FIXEDPITCH = 1,
  OP2_UNUSED = 2,
  OP2_DOUBLEVOICE = 4,
};
typedef struct opl_t opl_t;

struct opl_emu_registers
{
	uint16_t m_lfo_am_counter;
	uint16_t m_lfo_pm_counter;
	uint32_t m_noise_lfsr;
	uint8_t m_lfo_am;
	uint8_t m_regdata[OPL_EMU_REGISTERS_REGISTERS];
	uint16_t m_waveform[OPL_EMU_REGISTERS_WAVEFORMS][OPL_EMU_REGISTERS_WAVEFORM_LENGTH];
};

struct opl_emu_opdata_cache
{
	uint32_t phase_step;
	uint32_t total_level;
	uint32_t block_freq;
	int32_t detune;
	uint32_t multiple;
	uint32_t eg_sustain;
	uint8_t eg_rate[OPL_EMU_EG_STATES];
	uint8_t eg_shift;
};
struct opl_emu_fm_operator
{
	uint32_t m_choffs;
	uint32_t m_opoffs;
	uint32_t m_phase;
	uint16_t m_env_attenuation;
	enum opl_emu_envelope_state m_env_state;
	uint8_t m_key_state;
	uint8_t m_keyon_live;
	struct opl_emu_opdata_cache m_cache;
	struct opl_emu_registers* m_regs;
};
struct opl_emu_fm_channel
{
	uint32_t m_choffs;
	int16_t m_feedback[2];
	int16_t m_feedback_in;
};
typedef struct opl_timbre_t {
  unsigned long modulator_E862, carrier_E862;
  unsigned char modulator_40, carrier_40;
  unsigned char feedconn;
  signed char finetune;
  unsigned char notenum;
  signed short noteoffset;
} opl_timbre_t;
struct opl_emu_t
{
	uint32_t m_env_counter;
	uint8_t m_status;
	uint8_t m_timer_running[2];
	uint32_t m_active_channels;
	uint32_t m_modified_channels;
	uint32_t m_prepare_count;
	struct opl_emu_registers m_regs;
	struct opl_emu_fm_channel m_channel[OPL_EMU_REGISTERS_CHANNELS];
	struct opl_emu_fm_operator m_operator[OPL_EMU_REGISTERS_OPERATORS];
};
struct voicealloc_t {
  unsigned short priority;
  signed short timbreid;
  signed char channel;
  signed char note;
  unsigned char voiceindex;
};
struct opl_t {
  signed char notes2voices[16][128][2];
  unsigned short channelpitch[16];
  unsigned short channelvol[16];
  struct voicealloc_t voices2notes[18];
  unsigned char channelprog[16];
  int opl3;
  struct opl_emu_t opl_emu;
  struct opl_timbre_t opl_gmtimbres[ 256 ];
  struct opl_timbre_t opl_gmtimbres_voice2[ 256 ];
  int is_op2;
  enum op2_flags_t op2_flags[ 256 ];
};
static void opl_load_op2_voice(opl_timbre_t* timbre, uint8_t const* buff) ;
int opl_loadbank_op2(opl_t* opl, void const* data, int size ) ;
static void opl_load_op2_voice(opl_timbre_t* timbre, uint8_t const* buff) {
  timbre->modulator_E862 = buff[3];
  timbre->modulator_E862 <<= 8;
  timbre->modulator_E862 |= buff[2];
  timbre->modulator_E862 <<= 8;
  timbre->modulator_E862 |= buff[1];
  timbre->modulator_E862 <<= 8;
  timbre->modulator_E862 |= buff[0];
  timbre->carrier_E862 = buff[10];
  timbre->carrier_E862 <<= 8;
  timbre->carrier_E862 |= buff[9];
  timbre->carrier_E862 <<= 8;
  timbre->carrier_E862 |= buff[8];
  timbre->carrier_E862 <<= 8;
  timbre->carrier_E862 |= buff[7];
  timbre->modulator_40 = ( buff[5] & 0x3f ) | ( buff[4] & 0xc0 );
  timbre->carrier_40 = ( buff[12] & 0x3f ) | ( buff[11] & 0xc0 );
  timbre->feedconn = buff[6];
  timbre->finetune = 0;
  timbre->noteoffset = (int16_t)(buff[14] | ((uint16_t)buff[15] << 8));
}
int opl_loadbank_op2(opl_t* opl, void const* data, int size ) {
  if( size < 8 + 36 * 175 ) {
      return -3;
  }
  uint8_t const* buff = (uint8_t const*) data;
  int i;
  if ((buff[0] != '#') || (buff[1] != 'O') || (buff[2] != 'P') || (buff[3] != 'L') || (buff[4] != '_') || (buff[5] != 'I') || (buff[6] != 'I') || (buff[7] != '#')) {
    return(-3);
  }
  buff += 8;
  opl->is_op2 = 1;
  for (i = 0; i < 175; i++) {
    opl->op2_flags[i] = (enum op2_flags_t)( buff[0] | ((uint16_t)buff[1] << 8) );
    int finetune = buff[2];
    uint8_t fixednote = buff[3];
    buff += 4;
    opl_load_op2_voice(&opl->opl_gmtimbres[i], buff);
    opl->opl_gmtimbres[i].notenum = fixednote;
    buff += 16;
    opl_load_op2_voice(&opl->opl_gmtimbres_voice2[i], buff);
    opl->opl_gmtimbres_voice2[i].notenum = fixednote;
    opl->opl_gmtimbres_voice2[i].finetune += finetune - 128;
    buff += 16;
  }
  return(0);
}

File: flourine_reslut/fail/c/opl/opl_loadbank_op2/claude3/Hinted/opl_loadbank_op2.rs


use std::mem;

#[derive(Clone, Copy)]
#[repr(u8)]
enum OplEmuEnvelopeState {
    OplEmuEgAttack = 1,
    OplEmuEgDecay = 2,
    OplEmuEgSustain = 3,
    OplEmuEgRelease = 4,
    OplEmuEgStates = 6,
}

#[derive(Clone, Copy)]
struct Op2FlagsT(u8);

impl Op2FlagsT {
    const Op2Fixedpitch: Op2FlagsT = Op2FlagsT(1);
    const Op2Unused: Op2FlagsT = Op2FlagsT(2);
    const Op2Doublevoice: Op2FlagsT = Op2FlagsT(4);

    fn from_bytes(bytes: [u8; 2]) -> Self {
        let value = u16::from_ne_bytes(bytes);
        let value = value as u8;
        Op2FlagsT(value)
    }
}

const OPL_EMU_REGISTERS_OPERATORS: usize = OPL_EMU_REGISTERS_CHANNELS * 2;
const OPL_EMU_REGISTERS_WAVEFORMS: usize = 8;
const OPL_EMU_REGISTERS_CHANNELS: usize = 18;
const OPL_EMU_REGISTERS_REGISTERS: usize = 0x200;
const OPL_EMU_REGISTERS_WAVEFORM_LENGTH: usize = 0x400;

struct OplEmuRegisters {
    m_lfo_am_counter: u16,
    m_lfo_pm_counter: u16,
    m_noise_lfsr: u32,
    m_lfo_am: u8,
    m_regdata: Box<[u8; OPL_EMU_REGISTERS_REGISTERS]>,
    m_waveform: Box<[[u16; OPL_EMU_REGISTERS_WAVEFORM_LENGTH]; OPL_EMU_REGISTERS_WAVEFORMS]>,
}

struct OplEmuOpdataCache {
    phase_step: u32,
    total_level: u32,
    block_freq: u32,
    detune: i32,
    multiple: u32,
    eg_sustain: u32,
    eg_rate: Box<[u8; OplEmuEnvelopeState::OplEmuEgStates as usize]>,
    eg_shift: u8,
}

struct OplEmuFmOperator {
    m_choffs: u32,
    m_opoffs: u32,
    m_phase: u32,
    m_env_attenuation: u16,
    m_env_state: OplEmuEnvelopeState,
    m_key_state: u8,
    m_keyon_live: u8,
    m_cache: OplEmuOpdataCache,
}

struct OplEmuFmChannel {
    m_choffs: u32,
    m_feedback: Box<[i16; 2]>,
    m_feedback_in: i16,
}

struct OplTimbreT {
    modulator_E862: u32,
    carrier_E862: u32,
    modulator_40: u8,
    carrier_40: u8,
    feedconn: u8,
    finetune: i8,
    notenum: u8,
    noteoffset: i16,
}

struct OplEmuT {
    m_env_counter: u32,
    m_status: u8,
    m_timer_running: Box<[u8; 2]>,
    m_active_channels: u32,
    m_modified_channels: u32,
    m_prepare_count: u32,
    m_regs: OplEmuRegisters,
    m_channel: Box<[OplEmuFmChannel; OPL_EMU_REGISTERS_CHANNELS]>,
    m_operator: Box<[OplEmuFmOperator; OPL_EMU_REGISTERS_OPERATORS]>,
}

struct VoiceallocT {
    priority: u16,
    timbreid: i16,
    channel: i8,
    note: i8,
    voiceindex: u8,
}

struct OplT {
    notes2voices: Box<[[[i8; 2]; 128]; 16]>,
    channelpitch: Box<[u16; 16]>,
    channelvol: Box<[u16; 16]>,
    voices2notes: Box<[VoiceallocT; 18]>,
    channelprog: Box<[u8; 16]>,
    opl3: i32,
    opl_emu: OplEmuT,
    opl_gmtimbres: Box<[OplTimbreT; 256]>,
    opl_gmtimbres_voice2: Box<[OplTimbreT; 256]>,
    is_op2: i32,
    op2_flags: Box<[Op2FlagsT; 256]>,
}

fn opl_load_op2_voice(timbre: &mut OplTimbreT, buff: &[u8]) {
    timbre.modulator_E862 = u32::from(buff[3]); // wave select
    timbre.modulator_E862 = (timbre.modulator_E862 << 8) | u32::from(buff[2]); // sust/release
    timbre.modulator_E862 = (timbre.modulator_E862 << 8) | u32::from(buff[1]); // attack/decay
    timbre.modulator_E862 = (timbre.modulator_E862 << 8) | u32::from(buff[0]); // AM/VIB... flags

    timbre.carrier_E862 = u32::from(buff[10]); // wave select
    timbre.carrier_E862 = (timbre.carrier_E862 << 8) | u32::from(buff[9]); // sust/release
    timbre.carrier_E862 = (timbre.carrier_E862 << 8) | u32::from(buff[8]); // attack/decay
    timbre.carrier_E862 = (timbre.carrier_E862 << 8) | u32::from(buff[7]); // AM/VIB... flags

    timbre.modulator_40 = (buff[5] & 0x3f) | (buff[4] & 0xc0);
    timbre.carrier_40 = (buff[12] & 0x3f) | (buff[11] & 0xc0);

    timbre.feedconn = buff[6];
    timbre.finetune = 0;
    timbre.noteoffset = i16::from_le_bytes([buff[14], buff[15]]);
}

fn opl_loadbank_op2(opl: &mut OplT, data: &[u8], size: i32) -> i32 {
    if size < 8 + 36 * 175 {
        return -3;
    }

    let buff = data;

    if &buff[0..8] != b"#OPL_II#" {
        return -3;
    }

    opl.is_op2 = 1;

    for i in 0..175 {
        let op2_flags_bytes = [buff[i * 36 + 8], buff[i * 36 + 9]];
        let op2_flags = Op2FlagsT::from_bytes(op2_flags_bytes);
        let finetune = buff[i * 36 + 10] as i8;
        let fixednote = buff[i * 36 + 11];

        opl_load_op2_voice(&mut opl.opl_gmtimbres[i], &buff[i * 36 + 12..i * 36 + 28]);
        opl.opl_gmtimbres[i].notenum = fixednote;

        opl_load_op2_voice(&mut opl.opl_gmtimbres_voice2[i], &buff[i * 36 + 28..i * 36 + 44]);
        opl.opl_gmtimbres_voice2[i].notenum = fixednote;
        opl.opl_gmtimbres_voice2[i].finetune = opl.opl_gmtimbres_voice2[i].finetune.wrapping_add(finetune.wrapping_sub(128u8 as i8));

        opl.op2_flags[i] = op2_flags;
    }

    0
}

